addons/portal/models/mail_message.py

```diff
@@ -2,26 +2,99 @@
-            'message_type', 'subtype_id', 'is_internal', 'subject',  # message specific
-            'model', 'res_id', 'record_name',  # document related
-        ])
-
-    def _portal_message_format(self, fields_list):
-        vals_list = self._message_format(fields_list, legacy=True)
-        message_subtype_note_id = self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note')
-        IrAttachmentSudo = self.env['ir.attachment'].sudo()
-        for vals in vals_list:
-            vals['is_message_subtype_note'] = message_subtype_note_id and (vals.get('subtype_id') or [False])[0] == message_subtype_note_id
-            for attachment in vals.get('attachment_ids', []):
-                if not attachment.get('access_token'):
-                    attachment['access_token'] = IrAttachmentSudo.browse(attachment['id']).generate_access_token()[0]
+        return self._portal_message_format(
+            self._portal_get_default_format_properties_names(options=options)
+        )
+
+    def _portal_get_default_format_properties_names(self, options=None):
+        return {
+            'attachment_ids',
+            'author_avatar_url',
+            'author_id',
+            'body',
+            'date',
+            'id',
+            'is_internal',
+            'is_message_subtype_note',
+            'published_date_str',
+            'subtype_id',
+        }
+
+    def _portal_message_format(self, properties_names):
+        """ Format messages for portal frontend.
+
+        When asking for attachments: ensure an access token is present then
+        access them (using sudo).
+
+        :param set properties_names: set of fields / properties to compute
+
+        :return list: list of dict, one per message in self
+        """
+        message_to_attachments = {}
+        if 'attachment_ids' in properties_names:
+            properties_names.remove('attachment_ids')
+            self.sudo().attachment_ids.generate_access_token()
+            related_attachments = {
+                att_read_values['id']: att_read_values
+                for att_read_values in self.env['ir.attachment'].sudo().browse(
+                    self.sudo().attachment_ids.ids
+                ).read(
+                    ["access_token", "checksum", "id", "mimetype", "name", "res_id", "res_model"]
+                )
+            }
+            message_to_attachments = {
+                message.id: [
+                    self._portal_message_format_attachments(related_attachments[att_id])
+                    for att_id in message.attachment_ids.ids
+                ]
+                for message in self.sudo()
+            }
+
+
+        fnames = {
+            property_name for property_name in properties_names
+            if property_name in self._fields and property_name != 'id'
+        }
+        if fnames:
+            vals_list = self._read_format(fnames)
+        else:
+            vals_list = [{'id': message.id} for message in self]
```

Comment:
Wait, so we don't want the id if there are fields to format? I don't get it

Reply:
It was just a shortcut to avoid calling read_format if you don't want to read real fields, but maybe if you don't have anything to read probably better not call the method at all. Cleaned.